### PR TITLE
Make search failure cause rest failure

### LINKED ISSUE #16037: Reindex should copy status from any search failures
Currently it only copies from search failures. Oops.

### PR DESCRIPTION
**Pull request by nik9000**
Indexing failures have caused the reindex http request to fail for a while
now. Both search and indexing failures cause it to abort. But search
failures didn't cause a non-200 response code from the http api. This
fixes that.

Also slips in a fix to some infrequently failing rest tests.

Closes #16037

### CODE REVIEW

**Comment by clintongormley** [comment]
@nik9000 while you're on this, I think this issue would be of interest to you too: https://github.com/elastic/elasticsearch/issues/16555#issuecomment-183676958

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkIndexByScrollResponseContentListener.java [hunk 1]
```diff
@@ -41,6 +43,12 @@ protected RestStatus getStatus(R response) {
                 status = failure.getStatus();
             }
         }
+        for (ShardSearchFailure failure: response.getSearchFailures()) {
+            RestStatus failureStatus = ExceptionsHelper.status(failure.getCause());
+            if (failureStatus.getStatus() > status.getStatus()) {
```
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkIndexByScrollResponseContentListener.java]: This is kind of a weird way to order failures, but I guess it works. I think adding a comment might help others in the future in understanding your thinking

**Comment by dakrone** [comment]
LGTM left one minor comment

**Comment by nik9000** [comment]
Backported to 2.x with 7b1d567cd21c7dc3d7f2cd7fad889826125cd627

**Pull request merged**